### PR TITLE
Add cosn UFS to provide better ability to read and write cos

### LINKED ISSUE #12581: Add cosn UFS to provide better ability to read and write cos
**Is your feature request related to a problem? Please describe.**
COS has poor read and write performance and instability.Compared with COS, COSN is deeply optimized, and its read and write performance is doubled compared with COS, and it also brings better stability.

**Describe the solution you'd like**
Add CosN UFS 

**Describe alternatives you've considered**

**Urgency**
Will be helpful if this is taken on priority.

**Additional context**

### PR DESCRIPTION
**Pull request by jayzhenghan**
Fix #12581

### CODE REVIEW

**Comment by gpang** [comment]
@jayzhenghan Thanks! Could you add a PR description, including how this is different from the other `cos` scheme?

**Comment by jayzhenghan** [comment]
> @jayzhenghan Thanks! Could you add a PR description, including how this is different from the other `cos` scheme?

ok~,I am considering whether to directly replace cos or coexist with cos. first one, compatibility is smoother; second one is more secure but document the difference between cosn and cos and how to use it，compatibility is not good as first one.

**Comment by maobaolong** [comment]
@jayzhenghan Thanks for this contribution. You can create an issue label it with feature-request, and add description with motivation.

Then, you can insert `Fix #ISSUE_ID` to the start of PR description, so that when this PR merged, the related issue can be closed.

File: underfs/cosn/pom.xml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  <properties>
+    <build.path>${project.parent.parent.basedir}/build</build.path>
+    <hadoop.version>3.1.0</hadoop.version>
+  </properties>
+
+  <dependencies>
+    <dependency>
+      <groupId>com.qcloud.cos</groupId>
+      <artifactId>hadoop-cos</artifactId>
+      <version>3.1.0-5.8.5</version>
```
Comment (maobaolong) [inline underfs/cosn/pom.xml:36]: You can remove the `version` tag here, and maven can use the `hadoop-cos` definition from root pom.xml dependencyManagement.

File: underfs/cosn/pom.xml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    </dependency>
+    <dependency>
+      <groupId>com.qcloud</groupId>
+      <artifactId>cos_api-bundle</artifactId>
+      <version>5.6.19</version>
```
Comment (maobaolong) [inline underfs/cosn/pom.xml:41]: ditto

File: underfs/cosn/pom.xml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    <version>2.5.0-SNAPSHOT</version>
+  </parent>
+
+  <artifactId>alluxio-underfs-cosn</artifactId>
+  <name>Alluxio Under File System - Tencent Cloud cosn</name>
+  <description>Tencent Cloud cosn Under File System implementation</description>
+
+  <properties>
+    <build.path>${project.parent.parent.basedir}/build</build.path>
+    <hadoop.version>3.1.0</hadoop.version>
```
Comment (maobaolong) [inline underfs/cosn/pom.xml:29]: If you want to specify the `hadoop version` for this ufs, you can use another property name other than this `hadoop.version` to keep difference to alluxio main framework. For example, `cosn.hadoop.version`

**Review by maobaolong — CHANGES_REQUESTED** [review]
This PR looks like to create a new ufs module to make it possible to using cosn in Alluxio global namespace. I left some minor comments inline.

And, I noticed the `CosFileSystem` is an implementation of `Hadoop compatible filesystem`, so suggest to reference the https://www.youtube.com/watch?v=D65V3WpS5zQ  to reuse the `HdfsUnderFileSystem ` to avoid lots of redundant codes.

**Review by maobaolong — COMMENTED** [review]
`webui/npm-shrinkwrap.json`

Please revert this file changes

File: shaded/cosn/pom.xml [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.logging.log4j</groupId>
+      <artifactId>log4j-slf4j-impl</artifactId>
+      <scope>provided</scope>
+    </dependency>
+  </dependencies>
+
+  <profiles>
```
Comment (maobaolong) [inline shaded/cosn/pom.xml:56]: As now, we have only one profile, and it is active by default, you can put these dependencies directly, when you need to define another profile, we can bring this back.

**Comment by jayzhenghan** [comment]
> `webui/npm-shrinkwrap.json`
> 
> Please revert this file changes

Ok,I get it

**Review by maobaolong — CHANGES_REQUESTED** [review]
@jayzhenghan Thanks for this changes, it looks good overall, left some minor comments inline.

**Comment by jayzhenghan** [comment]
alluxio_rw_test

**Review by maobaolong — APPROVED** [review]
LGTM, thanks

**Comment by jayzhenghan** [comment]
> @jayzhenghan Thanks! Could you add a PR description, including how this is different from the other `cos` scheme?

Thank you for your suggestion，I have added the description in cos.md and cosn.md.

**Comment by jayzhenghan** [comment]
> @jayzhenghan Thanks! Could you add a PR description, including how this is different from the other `cos` scheme?

Thank you for your suggestion，I have added a description in cos.md and cosn.md.

**Comment by maobaolong** [comment]
@apc999 Could you please take another look?

**Review by maobaolong — APPROVED** [review]
LGTM, thanks for your contribution.

**Comment by jayzhenghan** [comment]
Fix #12581

**Comment by apc999** [comment]
@maobaolong , how about replace original COS module with this new Cosn module?

**Comment by jayzhenghan** [comment]
> @maobaolong , how about replace original COS module with this new Cosn module?

Taking into account the compatibility of cos, it is reserved, but maintenance will be stopped in version 2.6.0, and users are strongly recommended to use cosn

**Comment by maobaolong** [comment]
@apc999 I'm ok to replace the original COS module if there are no objection from users.

**Comment by gpang** [comment]
@jayzhenghan Thanks for this great contribution! Can you update the PR description with more details about this new/updated cos connector?

**Comment by jayzhenghan** [comment]
> @jayzhenghan Thanks for this great contribution! Can you update the PR description with more details about this new/updated cos connector?

@gpang @apc999 COS UFS has the problem of poor and unstable reader performance. In order to solve this problem, a file system like [COSN UFS] (COSN.html) has been introduced. Compared with COS, it has been deeply optimized, and its reading performance is doubled compared with COS, and it also brings better stability. For more usage, please refer to cos.md and cosn.md in pr.

**Review by apc999 — APPROVED** [review]

**Comment by apc999** [comment]
alluxio-bot, merge this please

**Comment by apc999** [comment]
thanks @jayzhenghan for the contribution!

**Pull request merged**